### PR TITLE
DHT indexing

### PR DESCRIPTION
**Pull request by maackle**

### CODE REVIEW

File: buntdbht.go [hunk 1]
```diff
@@ -47,6 +47,24 @@ func (ht *BuntHT) Open(options interface{}) (err error) {
...
+
+	// add additional indexes of type json to entries as specified in the index spec
+	// e.g. db.CreateIndex('tweet.text.body', "entry:*", buntdb.IndexJson("text.body"))
+	// indexing will then be automatically done on all future (and existing) keys that match
+
+	db := ht.db
+	for _, def := range indexSpec {
+		indexType := buntdb.IndexJSON(def.FieldPath)
+		if !def.Ascending {
+			indexType = buntdb.Desc(indexType)
```
Comment (maackle) [inline buntdbht.go:60]: Check for nonexistent field name

File: dht.go [hunk 2]
```diff
@@ -158,16 +160,71 @@ func NewDHT(h *Holochain) *DHT {
 	return &dht
 }
 
+type IndexDef struct {
+	ZomeName  string
+	IndexType string // Buntdb supported index types - int, float, string
```
Comment (maackle) [inline dht.go:165]: Can probz remove this

File: dht.go [hunk 3]
```diff
@@ -158,16 +160,71 @@ func NewDHT(h *Holochain) *DHT {
...
+}
+
+// get a single entries schema index specifications
+func indexSpecFromSchema(zomeName string, entryType string, schema string) IndexSpec {
+	var spec IndexSpec
+	indexFields := gjson.Get(schema, "indexFields").Array()
+	for _, field := range indexFields {
+		def := field.Map()
+		if len(def) != 1 {
+			panic("wtf!!")
```
Comment (maackle) [inline dht.go:198]: calm down

File: dht.go [hunk 4]
```diff
@@ -158,16 +160,71 @@ func NewDHT(h *Holochain) *DHT {
...
+}
+
 // Open sets up the DHTs data structures and store
 func (dht *DHT) Open(options interface{}) (err error) {
 	h := options.(*Holochain)
 	dht.h = h
 	dht.glog = &h.Config.Loggers.Gossip
 	dht.dlog = &h.Config.Loggers.DHT
-	dht.config = &h.Nucleus().DNA().DHTConfig
+	dht.config = &h.Nucleus().DNA().DHTConfig // BOOKMARK - use a similar method to get the index spec from app properties. Alternatively can put it in DHTConfig...
```
Comment (maackle) [inline dht.go:221]: note

**Comment by willemolding** [comment]
I'm going to go ahead and merge this to develop so we can use it at the hackathon

**Pull request merged**